### PR TITLE
Mountain fold

### PR DESCRIPTION
**Pull request by ptpham**
- Please don't mind currently sparse README -- we can flesh it out later. =P
- Having the var in a for-of is important. Without it, the variable declaration will leak out of scope.

### CODE REVIEW

File: paper.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    var src = halfedge.src, dst = halfedge.dst;
+    var intersect = geometry.intersectSegment(vertex, direction, src, dst);
+    if (intersect && v3.distance(src, intersect) > EPSILON) {
+      if (v3.distance(dst, intersect) > EPSILON) this.split(halfedge, intersect);
+      intersects.push(halfedge);
+    }
+  }
+
+  // Impose an ordering on the halfedges and connect halfedges in the same face
+  var clip = this.sheet.ordering.impose(intersects);
```
Comment (cixzhang) [inline paper.js]: I'm not quite sure what's going on in this section. Shouldn't ordering related operations be handled by Sheet? It seems like a bit of a reach.

File: sheet.js [hunk 2]
```diff
@@ -32,13 +44,22 @@ Sheet.prototype.seamCompare = function(seam) {
...
-  this.ordering.inject(position, extracted);
+  this.ordering.inject(position, extracted, upwards ? 1 : 0);
+  return extracted;
+};
+
+Sheet.prototype.foldDown = function(seam) {
+  return this.basicFold(seam, false);
+};
+
+Sheet.prototype.foldUp = function(seam) {
```
Comment (cixzhang) [inline sheet.js]: It should work, but can we add a test for this?

File: paper.js (hunk 1 above)
Comment (cixzhang) [inline paper.js]: Okay, I'm reading it again and I think I understand it a bit better. Above, we find halfedges that intersect with the line. Here we're trying to connect the ones that are on the same face.

Um, could we iron out language a bit? @__@ I get confused and have to go through the source each time I read 'impose' or 'connect' or 'slice'.

File: paper.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  for (var slice of clip) {
+    var sorted = _.sortBy(slice, function(halfedge) {
+      return v3.distance(halfedge.dst, vertex);
+    });
+    for (var i = 1; i < sorted.length; i++) {
+      var previous = sorted[i - 1];
+      var current = sorted[i];
+
+      if (_.contains(current.face, previous)) {
```
Comment (cixzhang) [inline paper.js]: Seems potentially dangerous to make this check all the way out here, several abstractions away from Halfedge. Also, do we ever need to worry about irregular faces? (i.e. Is it possible to have concave faces?)

File: sheet.js [hunk 4]
```diff
@@ -32,13 +44,22 @@ Sheet.prototype.seamCompare = function(seam) {
   return this.ordering.compare(primary, mirror);
 };
 
-Sheet.prototype.foldDown = function(seam) {
+Sheet.prototype.basicFold = function(seam, upwards) {
   var compare = this.seamCompare(seam);
   var affected = this.seamFlood(seam);
   var extracted = this.ordering.extract(affected);
   if (compare == 0) extracted.reverse();
   var position = _.chain(seam).pluck('mirror').min(this.ordering.iteratee).value();
```
Comment (cixzhang) [inline sheet.js]: Should this still be the min for foldUp?

File: twgl.js [hunk 5]
```diff
@@ -16,7 +16,9 @@ twgl.v3.create = function(x, y, z) {
   if (y) result[1] = y;
   if (z) result[2] = z;
   return result;
-}
+};
+
+twgl.v3.distance = function(a, b) { return v3.length(v3.subtract(a, b)); };
```
Comment (cixzhang) [inline twgl.js]: Why not geometry?

**Comment by ptpham** [comment]
Thanks for the feedback! Earliest time I can address them is probably Sunday morning. XD

File: sheet.js [hunk 6]
```diff
@@ -12,6 +12,18 @@ function Sheet() {
   this.ordering.merge(halfedge, halfedge.mirror);
 }
 
+Object.defineProperty(Sheet.prototype, 'interior', {
+  get: function() { return this.seamFlood([this.initial]); }
```
Comment (cixzhang) [inline sheet.js]: I wonder if we can (or should) resolve this with ```this.inside``` which serves sort of the same purpose.

Comment (ptpham) [inline sheet.js]: I'm thinking that it actually doesn't make that much sense to merge interior with inside. Inside is just supposed to be a cache to answer a boolean query whereas interior allows iteration over elements ...

File: twgl.js (hunk 5 above)
Comment (ptpham) [inline twgl.js]: Addressed.

File: paper.js (hunk 3 above)
Comment (ptpham) [inline paper.js]: I've just added a comment that we don't handle non-convex faces. XD

File: sheet.js (hunk 4 above)
Comment (ptpham) [inline sheet.js]: Addressed.

File: sheet.js (hunk 2 above)
Comment (ptpham) [inline sheet.js]: Addressed.

**Comment by cixzhang** [comment]
Looks good :D

**Pull request merged**